### ISSUE TITLE
Copy refactor

### ISSUE THREAD

**Issue by j7b** [jackc/pgx#252]
I started hacking in a CopyOut function and accidentally ended up refactoring copy in general.

Ironically the binary part isn't done, basically format binary will require a copyfromsource-type but I'm thinking it should be refactored into BinaryReader and BinaryWriter with implementations that would be basically a buffer (the [][]interface{} type thing) and probably typed wrappers for the csv reader and writer. The query form of a copyout isn't done yet either.

It's rough and incomplete but works well enough now to prove the concept, tho it's probably better to just take my word for it at this point.

Basically it works like:
```
tx,_ := db.Begin()
copier := new(Copier)
// or copier,err := NewCopier(CSVFormat,CopierHeader(true))
buf := new(bytes.Buffer)
err := copier.Copy(tx).Table(Identity{`my`,`table`}).To(buf) // or .From(buf)
err = copier.Copy(tx).Query(`select 'I''m pretty sure there''s a way to get unnamed 
in a copy without materializing it' || $1`,"!").To(buf) // soon...
```

Blob to date at https://github.com/j7b/pgx/blob/c11cec9d86b494905eadc3580de7d6a06267d245/copy_redux.go

@jackc this headed in an ok direction?

**Comment by jackc** [comment]
Can you help me understand the high level goals? Using `CopyFrom` to load data into the server can offer better performance than `Exec("insert ...")`. Are we hoping for a similar gain with a `COPY TO` vs `Query("select ...")`. Or is this more for convenience to generate CSV from server or the like?

As far as the API goes, it's hard to say much without understanding the desired result, but that said hear are my initial thoughts.

I'd like to avoid any further changes to the copy API in `v2` (especially given the recent addition of `CopyFrom` and deprecation of `CopyTo`), so if these changes work out they would probably land in `v3` (I hope for a `v3` release in the next couple months).

CSV or text format protocol can be tricky to handle. At the least `DateStyle` and `IntervalStyle` affect input and output. Possibly quoting logic would be affected by server settings as well. The current implementation of `CopyFrom` sidesteps that issue by only supporting binary format.

In the last example with `Copy(tx).Query`, it looks like it is using a query argument. `COPY` cannot be used with prepared statements, and internally pgx always uses prepared statements to safely handle arguments. At the moment that would be a problem. I'm _considering_ adding client side argument sanitization to `v3` to support queries without prepared statements (necessary to resolve #172), but that would need to be handled with great care.

**Comment by j7b** [comment]
The main goal is convenience, use cases where you have a reader or writer and the path of least resistance is a statement represented as a reader or writer. There's also a capacity planning aspect, in environment where clients are known to be resource-constrained and backends are known to be resource-rich letting the backend do all the heavy lifting and minimizing operations and allocations on the client is desirable.

The design is intended to be completely agnostic about what postgresql.conf says and what you've done to your session.  If a client is choosing to marshal/unmarshal csv fields, for example, that's their responsibility. Dunno if other people do this often but I'm pretty keen on copying binary for cases where clients access multiple backends and copying to_jsons and json_aggs without the intercession of scans for garden-variety web API's and such.

I've actually been thinking about types in general quite a bit in thinking about BinaryReader and BinaryWriter, because preferably it should be driver-agnostic. I like the idea of everything defined in native wire format and types that are reasonable candidates would have MarshalPG/UnmarshalPG/IsNull methods with the intent things look along the lines of:

```
buf := new(BinaryBuffer)
if err := copier.Copy(tx).Query(`select 'stuff'`).To(buf) ; err == nil {
   for _,scanners := range buf {
       err = scanner.Scan(&unmarshalbletype,&wiretype)
     // ....
   }
}

```

Basically I think it's possible for the binary format to have general utility similar to gobs or protocol buffers in a Go implementation, and there's decent value in such an implementation.

As to the sanitizing problem I'm pretty sure I've seen a libpq or possibly Java implementation that solved the problem with a cursor. I also have some some doblock ideas that would probably work pretty well. I also like the idea of some sort of RawQuery type intended for cases where you don't need sanitizing, but the implementation would probably have to be very specific and somewhere between a little and very ugly, but that's probably ok and desirable.

In any event binary format is semi-blocked on design and test of a binaryreadwriter, tho it has utility without it, and from's not easy to use without q format, so there's a fair amount of work before a pr'd be forthcoming.

**Comment by jackc** [comment]
I think this issue is obsolete now. Things have improved quite a bit from `v2` to `v4`.
